core/src/storage2/collections/binary_heap/children.rs

```diff
@@ -0,0 +1,209 @@
+/// stored in one single storage cell. This reduces storage access operations
+/// of the binary heap algorithm.
+#[cfg_attr(feature = "std", derive(TypeInfo))]
+#[derive(scale::Encode, scale::Decode, Default, PartialEq, Eq, Debug)]
+pub struct Children<T: PackedLayout + Ord> {
+    left: Option<T>,
+    right: Option<T>,
+}
+
+/// The position which a child has in a `Children` object.
+#[derive(PartialEq, Debug)]
+pub enum ChildPosition {
+    Left,
+    Right,
+}
+
+/// Number of elements stored in each `Children` object.
+///
+/// Note that the first `Children` object (at index `0`) will only ever
+/// contain one element (the root element).
+pub(crate) const CHILDREN_PER_NODE: u32 = 2;
+
+/// Returns the index of the `Children` object in which the `n`-th element of
+/// the heap is stored.
+pub(crate) fn get_children_storage_index(n: u32) -> u32 {
+    if n == 0 {
+        return 0
+    }
+    // The first `Children` object only ever contains the root element:
+    // `[Some(root), None]`. So when calculating indices we need to account
+    // for the items which have been left empty in the first `Children` object.
+    let padding = CHILDREN_PER_NODE - 1;
+    (n + padding) / CHILDREN_PER_NODE
+}
+
+/// Returns the `ChildPosition` of the `n`-th heap element.
+///
+/// For example, the element `3` is found at the child position `0`
+/// (within the `Children` object at storage index `2`).
+pub(crate) fn get_child_pos(n: u32) -> ChildPosition {
+    let storage_index = get_children_storage_index(n);
+    match (storage_index, n) {
+        (0, 0) => ChildPosition::Left,
+        (0, _) => panic!("first children object contains only the root element"),
+        (_, _) => {
+            let child_pos = (n - 1) % CHILDREN_PER_NODE;
+            match child_pos {
+                0 => ChildPosition::Left,
+                1 => ChildPosition::Right,
+                _ => {
+                    unreachable!(
+                        "CHILDREN_PER_NODE is 2, following the modulo op index must be 0 or 1"
+                    )
+                }
+            }
+        }
+    }
+}
+
+impl<T> Children<T>
+where
+    T: PackedLayout + Ord,
+{
+    /// Creates a new `Children` object with a left and right element.
+    pub fn new(left: Option<T>, right: Option<T>) -> Self {
+        Self { left, right }
+    }
+
+    /// Returns `true` if no child exists in this object, otherwise `false`.
+    pub fn is_empty(&self) -> bool {
+        self.left.is_none() && self.right.is_none()
+    }
+
+    /// Returns a shared reference to the element at `which`.
+    pub fn child(&self, which: ChildPosition) -> Option<&T> {
+        match which {
+            ChildPosition::Left => self.left.as_ref(),
+            ChildPosition::Right => self.right.as_ref(),
+        }
+    }
+
+    /// Returns an exclusive reference to the element at `which`.
+    pub fn child_mut(&mut self, which: ChildPosition) -> &mut Option<T> {
```

Comment:
Why this returns `&mut Option<T>` whereas `child` returns `Option<&T>`?
Both APIs should behave the same! `Option<&mut T>` is preferable over `&mut Option<T>` if possible.